src/Microsoft.Health.Dicom.Core/Features/Workitem/WorkitemQueryTagService.cs

```diff
@@ -5,57 +5,63 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using EnsureThat;
+using FellowOakDicom;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Dicom.Core.Exceptions;
+using Microsoft.Health.Dicom.Core.Features.Common;
+using Microsoft.Health.Dicom.Core.Features.ExtendedQueryTag;
 using Microsoft.Health.Dicom.Features.Common;
 
 namespace Microsoft.Health.Dicom.Core.Features.Workitem
 {
-    public class WorkitemQueryTagService : IWorkitemQueryTagService, IDisposable
+    public sealed class WorkitemQueryTagService : IWorkitemQueryTagService, IDisposable
     {
         private readonly IIndexWorkitemStore _indexWorkitemStore;
-        private readonly AsyncCache<IReadOnlyCollection<WorkitemQueryTagStoreEntry>> _queryTagCache;
-        private bool _disposed;
+        private readonly AsyncCache<IReadOnlyCollection<QueryTag>> _queryTagCache;
+        private readonly IDicomTagParser _dicomTagParser;
+        private readonly ILogger _logger;
 
-        public WorkitemQueryTagService(IIndexWorkitemStore indexWorkitemStore)
+        public WorkitemQueryTagService(IIndexWorkitemStore indexWorkitemStore, IDicomTagParser dicomTagParser, ILogger<WorkitemQueryTagService> logger)
         {
             _indexWorkitemStore = EnsureArg.IsNotNull(indexWorkitemStore, nameof(indexWorkitemStore));
-            _queryTagCache = new AsyncCache<IReadOnlyCollection<WorkitemQueryTagStoreEntry>>(ResolveQueryTagsAsync);
+            _queryTagCache = new AsyncCache<IReadOnlyCollection<QueryTag>>(ResolveQueryTagsAsync);
+            _dicomTagParser = EnsureArg.IsNotNull(dicomTagParser, nameof(dicomTagParser));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
         }
 
         public void Dispose()
         {
-            Dispose(disposing: true);
+            _queryTagCache.Dispose();
             GC.SuppressFinalize(this);
         }
 
-        protected virtual void Dispose(bool disposing)
+        public async Task<IReadOnlyCollection<QueryTag>> GetQueryTagsAsync(CancellationToken cancellationToken = default)
         {
-            if (!_disposed)
-            {
-                if (disposing)
-                {
-                    _queryTagCache.Dispose();
-                }
-
-                _disposed = true;
-            }
+            return await _queryTagCache.GetAsync(cancellationToken: cancellationToken);
         }
 
-        public async Task<IReadOnlyCollection<WorkitemQueryTagStoreEntry>> GetQueryTagsAsync(CancellationToken cancellationToken = default)
+        private async Task<IReadOnlyCollection<QueryTag>> ResolveQueryTagsAsync(CancellationToken cancellationToken)
         {
```

Comment:
See if there logs, both info or perf once you want to track to help debug easily.

Reply:
Sure